05.wc/wc.rb

```diff
@@ -0,0 +1,96 @@
+#!/usr/bin/env ruby
+
+# frozen_string_literal: true
+
+require 'optparse'
+
+NO_OPTION_LENGTH = 7
+
+def main
+  params = option_params
+  File.pipe?($stdin) ? print_type_pipe(params) : print_type_file(params)
+end
+
+def option_params
+  opt = OptionParser.new
+  params = {}
+  opt.on('-l') { |v| params[:l] = v }
+  opt.on('-w') { |v| params[:w] = v }
+  opt.on('-c') { |v| params[:c] = v }
+  opt.parse!(ARGV)
+  params
+end
+
+def print_type_pipe(params)
+  counts = get_counts($stdin.read)
+  length = single_option?(params) ? 0 : NO_OPTION_LENGTH
+  puts join_counts(counts, length, params)
+end
+
+def print_type_file(params)
+  counts_array = ARGV.map do |file_name|
+    get_counts(File.read(file_name), file_name)
+  end
+
+  length = 0
+  total_counts = sum_counts(counts_array)
+  unless single_option?(params) && single_file?
+    length = total_counts.map do |key, value|
+      key != :file_name ? value : 0
+    end.max.to_s.length
+  end
+
+  counts_array.each do |counts|
+    puts join_counts(counts, length, params)
+  end
+
+  return if single_file?
+
+  puts join_counts(total_counts, length, params)
+end
+
+def get_counts(text, file_name = '')
+  {
+    line: get_return_count(text),
+    word: get_words_count(text),
+    size: text.bytesize,
+    file_name:
+  }
+end
+
+def single_option?(params)
+  params.length == 1
+end
+
+def single_file?
+  ARGV.length == 1
+end
+
+def join_counts(counts, length, params)
+  list = []
+  list << counts[:line].to_s.rjust(length) if params[:l] || params.empty?
+  list << counts[:word].to_s.rjust(length) if params[:w] || params.empty?
+  list << counts[:size].to_s.rjust(length) if params[:c] || params.empty?
```

Comment:
```suggestion
  list = %i(line word size).map do |key|
    counts[key].to_s.rjust(length) if params[key] || params.empty?
  end
```

paramsのkeyもcountsと同じにするとこんな書き方ができそうです

Reply:
`:line ~ :size` に関しては「シンボル」のみが変わるだけでしたので、ご指摘の通りに `map` で処理をするようにしました。

また、今までは「条件に当てはまった場合のみ list へ格納される」仕組みでしたが、上記の場合は `wc.rb -l` といった指定の場合には `nil` が含まれた配列が作成されるため、`reject` で弾くようにしました。

その他、`params[key]` には `-l` のようなパラメータが入る状態でしたので、`option_params` メソッドの以下の部分を変更して `:line` などの後に使用する名称に合わせました。

```ruby
  opt.on('-l') { |v| params[:line] = v }
  opt.on('-w') { |v| params[:word] = v }
  opt.on('-c') { |v| params[:size] = v }
```